File: problems/34_find-first-and-last-position-of-element-in-sorted-array/main.cpp
#include <iostream>
#include <utility>
#include <vector>

class Solution
{
public:
    std::vector<int> searchRange(std::vector<int> &nums, int target)
    {
        auto l = lower_bound(nums.begin(), nums.end(), target);
        auto r = upper_bound(nums.begin(), nums.end(), target);
        if (l == nums.end() || *l != target)
            return {-1, -1};
        return {(int)(l - nums.begin()), (int)(r - nums.begin() - 1)};
    }
};

int main()
{
    Solution solution;
    std::vector<std::tuple<std::vector<int>, int, std::vector<int>>> samples = {
        {{5, 7, 7, 8, 8, 10}, 8, {3, 4}},
        {{5, 7, 7, 8, 8, 10}, 6, {-1, -1}},
        {{}, 0, {-1, -1}},
    };

    for (const auto &[nums, target, expected] : samples)
    {
        auto input = nums;
        std::vector<int> result = solution.searchRange(input, target);

        std::cout << "input: nums = [";
        for (size_t i = 0; i < nums.size(); ++i)
        {
            std::cout << nums[i] << (i + 1 < nums.size() ? "," : "");
        }
        std::cout << "], target = " << target
                  << ", placeholder result: [" << result[0] << "," << result[1] << "]"
                  << ", expected: [" << expected[0] << "," << expected[1] << "]\n";
    }

    return 0;
}

File: problems/34_find-first-and-last-position-of-element-in-sorted-array/main.rs
struct Solution;

impl Solution {
    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
        fn lower_bound(nums: &Vec<i32>, target: i32) -> usize {
            let mut left = 0;
            let mut right = nums.len();
            while left < right {
                let mid = left + (right - left) / 2;
                if nums[mid] < target {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }
            left
        }

        fn upper_bound(nums: &Vec<i32>, target: i32) -> usize {
            let mut left = 0;
            let mut right = nums.len();
            while left < right {
                let mid = left + (right - left) / 2;
                if nums[mid] <= target {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }
            left
        }

        let l = lower_bound(&nums, target);
        let r = upper_bound(&nums, target);
        if l < r && nums[l] == target {
            vec![l as i32, (r - 1) as i32]
        } else {
            vec![-1, -1]
        }
    }
}

fn main() {
    let samples: Vec<(Vec<i32>, i32, Vec<i32>)> = vec![
        (vec![5, 7, 7, 8, 8, 10], 8, vec![3, 4]),
        (vec![5, 7, 7, 8, 8, 10], 6, vec![-1, -1]),
        (vec![], 0, vec![-1, -1]),
    ];

    for (nums, target, expected) in samples {
        let result = Solution::search_range(nums.clone(), target);
        println!(
            "input: nums = {:?}, target = {}, placeholder result: {:?}, expected: {:?}",
            nums, target, result, expected
        );
    }
}
